GIT-7: Private Repository has no tests due to which the build fails.

Haxsen opened the pull request:
Hi, please check.
I made a common function (just like #3 ) for get public/private repository methods.
It might have mistakes, so let me know.
Tested on Ubuntu, no coverage errors anymore. But please verify.
Thanks.

Haxsen committed:
GIT-7: Private Repository has no tests due to which the build fails.

Add tests for the new getPrivateRepository method.

Haxsen committed:
GIT-7: Private Repository has no tests due to which the build fails.

Add tests for the new getPrivateRepository method.

Haxsen commented:
@vmassol  Hi, can you check this? Thanks.

vmassol commented on src/test/java/org/xwiki/git/script/GitScriptServiceTest.java:
```diff
@@ -55,6 +56,54 @@
 
     private File testRepository;
 
+    @Unstable
```
Same error as Usama... I don't understand why you both adding this annotation. Could you explain?

Where did you find the definition of it and the doc about it that would suggest to add it here? I'd like to see if there's something to improve.

Thx

Haxsen replied:
Hm, I thought we add this with any new method introduced (test methods as well). But I see now we dont need it for tests.

vmassol replied:
Yes but why did you think we would need unstable? What is the purpose of it? you need to understand everything you do! And if something seems weird you need to question it. That’s a good way to progress.

Haxsen replied:
We use `@Unstable` notation to show in the backward-compatibility reports that this is a new API which will change quite frequently (as I read on Dev Practices). We use it for new methods or APIs. Now why I shouldn't have used it here is because it's not a Public API (which I forgot to remember from the Dev Practices) and also because tests are not the new API itself but to just test them. There's a lot of stuff to remember and I try as much as I can.

vmassol replied:
Yes there's a lot! Indeed a test is not a public user API. We use @Unstable for public APIs only (not just for APIs). This allow us to stabilize the API and not break backward compatibility since user must use unstable public API at their own risks.

There's a lot but this practice of handling backward compatibility is really important and novel (you won't learn it elsewhere). So if you fully understand how we do this, you'll have learnt a lot more than to learn how to code in java or write unit tests since all that you'll learn anyway and it' pretty basic and obvious :) I highly recommend that you spend some time understand that including how we use aspects for it.

Haxsen replied:
Yep @vmassol this is actually giving me immensely useful knowledge which I never had in the past (by programming for a few years in different languages). I am not being very grateful I know, but it has been a steep learning curve for me. I admit that my reading and memory ability is not very good but I will try to gain your trust :)
And just to let you know, I've watched your speech on youtube about revapi, mutation testing (and more), but I only read the slides since I don't understand french. Also I saw your slides on slideshare about coverage testing. So my point is, I am not just trying to find something to copy paste as most students do. I understand what I can and always I will be glad to correct myself. I don't mean I unlike your criticism, I do appreciate your criticism as it motivates me to find new things.

vmassol replied:
Thanks for your message @Haxsen. Just want to say that while I can be tough, there's really nothing personal and I really appreciate you sticking with me. TBH I'm sure I'm not doing the best mentoring job I could. I'm a bit short on time and I'm sure that I don't spend enough time explaining things, and there's a lot of things. They're all written down on dev.xwiki.org but there's really a lot. It probably takes about 6 months to understand them all so it's understandable that you cannot grasp everything at once.

I appreciate you checking out the videos. Some more info on this:
* You can check all talks and presentations I gave on my blog  (https://massol.net), at https://massol.myxwiki.org/xwiki/bin/view/Activities/Data/
* There's also a list of talks at https://dev.xwiki.org/xwiki/bin/view/PresentationIdea/. For ex related to revapi and mutation, in English, the last one is at https://massol.myxwiki.org/xwiki/bin/view/Activities/Data/BruJUG2020/. There was no recording though. Now for a recording, there's the FOSDEM one at https://massol.myxwiki.org/xwiki/bin/view/Activities/Data/FOSDEM2018JavaTestingNext/

Thanks!

Haxsen replied:
That's great, thanks :)

Haxsen replied:
BTW All of those lectures are extremely helpful, specially the BruJUG2020. I never had such knowledge.
Thanks a lot and keep them going :)

vmassol commented on src/test/java/org/xwiki/git/script/GitScriptServiceTest.java:
```diff
@@ -55,6 +56,54 @@
     private File testRepository;
 
+    @Unstable
+    private void getGitRepositoryAndFindAuthors(String username, String accessCode) throws Exception
```
How can a method named "get*" not return anything? :)

vmassol replied:
Normally you're going to say that this is how the test methods are named :) That's because we're following the test best practices defined at https://dev.xwiki.org/xwiki/bin/view/Community/Testing/JavaUnitTesting/#HBestpractices

Haxsen replied:
I saw on the original code it had "get*" on the void so I created just like that.

vmassol replied:
Again there was a reason to have get in a @Test. You need to ask yourself why and understand it. Do you understand why now?

Haxsen replied:
From what I perceive: the get* here is not in the same meaning of returning something but rather to test the use case (which is getting that repository from our created Test repo in the `@Before` method).

Haxsen replied:
Please let me know if I am wrong about this.

vmassol commented on src/test/java/org/xwiki/git/script/GitScriptServiceTest.java:
```diff
@@ -55,6 +56,54 @@
+        if (username == null && accessCode == null) {
+            repository = service.getRepository(this.testRepository.getAbsolutePath(), TEST_REPO_CLONED);
+        } else {
+            repository = service.getPrivateRepository(this.testRepository.getAbsolutePath(), TEST_REPO_CLONED,
```
I think I would name the 2 APIs with the same method name and just change the parameters. Seems more consistent to me. `service.getRepository(... username, accessCode)`.

vmassol replied:
This would also allow you use simply:

```
repository = service.getRepository(this.testRepository.getAbsolutePath(), TEST_REPO_CLONED, username, accessCode);
```

for both cases. And you'd check for null in the implementation class.

Haxsen replied:
You mean overloading methods as I was creating in the #2 ?

vmassol replied:
Yes

Haxsen replied:
Doing it :)

vmassol commented:
I don't see where you create a private repository, ie where you validate that what you did worked? It seems to me that all that is checked is that getPrivateRepository() is called. But if getPrivateRepository()'s implementation was replaced to simply call getRepository() the test would still pass, right? ie the test is not testing much and I fail to see its interest.

Haxsen commented:
Ok @vmassol I just read your comments *late* now. I will update you tomorrow on this.
Thanks for the review.

Haxsen commented:
@vmassol , So far I have been searching how I can test this. A way to access a private repo is to use an SSH deploy key in the repository settings but after using this method, I just realized that I am not actually testing the method which needs `username` and `accessCode` of the git user instead of the ssh key. So, another way is to use my username 'Haxsen' and the token I generated in my settings. But I am unsure of my account's security this way.
What would you recommend?

vmassol commented:
I don’t know as I haven’t researched it. All I know is that you shouldn’t depend on anything remote. The test must be self sufficient and should work fine locally and in offline mode.

Now there’s no need to do an integration test here since it’s not our goal to test the jgit code (it’s tested on that project!). All we need to test is that we call the right API on it.

Haxsen commented:
> All we need to test is that we call the right API on it

You mean just call it, not check the result? I am confused.

Haxsen committed:
GIT-7: Private Repository has no tests due to which the build fails.

Remove @Unstable.

Haxsen committed:
GIT-7: Private Repository has no tests due to which the build fails.

Change getPrivateRepository to getRepository.

Haxsen committed:
GIT-7: Private Repository has no tests due to which the build fails.

Change getPrivateRepository to getRepository (overload)

Haxsen committed:
GIT-7: Private Repository has no tests due to which the build fails.

Change getPrivateRepository to getRepository (overload)

vmassol commented on src/test/java/org/xwiki/git/script/GitScriptServiceTest.java:
```diff
@@ -54,6 +54,52 @@
+        } else {
+            repository = service.getRepository(this.testRepository.getAbsolutePath(), TEST_REPO_CLONED,
+                username, accessCode);
+        }
```
You haven't changed this yet. You don't need the "if" here anymore (see my previous message). Thx

Haxsen replied:
I tried that and ran into coverage errors during maven build. (The test was only testing the new overloaded method)

Haxsen replied:
See: ![vmplayer_KClfMISxpB](https://user-images.githubusercontent.com/33906649/79054959-a845b000-7c62-11ea-949c-f752cf62ece4.png)

Haxsen replied:
From what I remember from googling, I believe this is related META-INF..
just found it here again: [link](https://github.com/unitsofmeasurement/unit-api/issues/161)
Let me try some solutions with this.

Haxsen replied:
Actually, no, the error is different now (built again).
Now it's 0.86 instead of 0.0 (and fails because less than 0.87):
![image](https://user-images.githubusercontent.com/33906649/79055125-e42d4500-7c63-11ea-95bd-17308fa4daa2.png)

Haxsen replied:
Ok now I know, doing the changes and trying again.

Haxsen replied:
Hm still same problem. (ratio is 0.86)
What I did: I just added `username, accessCode` in the parameters.
Like:
```
repository = service.getRepository(this.testRepository.getAbsolutePath(), TEST_REPO_CLONED, "test user", "TestAccessCode");
```
Whereas the whole java file is unchanged.

Haxsen replied:
Changes commited so you can see the code.

vmassol commented:
Could you explain what you've changed for the test? Thx

Haxsen commented:
Nothing, just removed Unstable notation.
(and renamed `getPrivateRepository` to `getRepository` in all files).
Edit: I did extra changes (only calling the new overloaded method). So now getting 0.86 coverage. I am still waiting for your reply.

Haxsen committed:
GIT-7: Private Repository has no tests due to which the build fails.

Remove `if` block .

Haxsen committed:
GIT-7: Private Repository has no tests due to which the build fails.

Remove `if` block for the other function.

Haxsen committed:
GIT-7: Private Repository has no tests due to which the build fails.

Remove params.

Haxsen committed:
GIT-7: Private Repository has no tests due to which the build fails.

Remove extra indent.

Haxsen commented:
Hi, I am getting coverage error (ratio is 0.86 instead of 0.87). Possibly I may have not done it right.
I've updated the code on this PR so you can see.

Haxsen commented:
Hi @vmassol, I hope you're fine. If you're free can you check my replies? Thanks.

vmassol commented:
@Haxsen Thanks. Here's what I did:

* Fixed your `@since` tags which were wrong
* Add a missing test. That was the important part that I had asked you to add. Actually this should have been the most important part of the PR since it required to be creative and to find a solution. Note that I didn't know how to do it so I had to research it.
* Fixed a few small details.
* Increased the test coverage from 87% to 91% thanks to the new tests.
* Closed GIT-7 which was wrongly created. Reopened GIT-6 and committed the fix against it.

Thanks

Haxsen commented:
Thanks @vmassol for the summary. I am just looking at your code, it's absolutely amazing. I will learn from this as much as I can.
Thank you again for your valuable time!
